---
title: "The evolution of diverse antimicrobial responses in
vancomycin-intermediate Staphylococcus aureus and its therapeutic implications"
author: Kyle Card and Arda Durmaz
Date: 1/8/2025
output: html_notebook
---

## Prerequisites

Clear current environment

```{r Clear current environment, include = FALSE}
rm(list = ls())
```
<br>

Load packages

```{r Packages, message = FALSE}
library(tidyverse)
library(magick)
library(pdftools)
library(grid)
library(cowplot)
library(proxy)
library(foreach)
library(doParallel)
library(glmnet)
library(pheatmap)
```
<br>

```{r Suppress summarise info}
options(dplyr.summarise.inform = FALSE)
```
<br>

***

# Experimental evolution under vancomycin selection

Load data

```{r Vancomycin MIC data}
# Data report the minimum inhibitory concentrations (MICs) of the 18
# experimental populations evolved under increasing vancomycin concentrations
# until they reach intermediate resistance levels (4 - 8 ug/mL).
vancomycin_MICs <- read_csv("data/vancomycin_MICs.csv",
  show_col_types = FALSE
)
```
<br>

```{r Function that plots vancomycin MIC data, messages = FALSE}
generate_vanc_plot <- function(dat, combined = FALSE) {
  # In the MIC column, replace all "-" with "NA"
  dat$MIC <- dat$MIC %>%
    str_replace_all("-", "NA") %>%
    as.numeric()

  # Remove all rows with NA values in the MIC column
  dat <- dat %>%
    filter(!is.na(MIC))

  # Plot of vancomycin MICs over time
  vanc_plot <- dat %>%
    ggplot(aes(
      x = day,
      y = MIC
    )) +
    geom_rect(
      aes(
        xmin = -Inf,
        xmax = Inf,
        ymin = 4,
        ymax = 8
      ),
      fill = "#E1EEF4",
      alpha = 0.1
    ) +
    geom_line(linewidth = 1.2) +
    geom_point(size = 2) +
    facet_wrap(~population,
      nrow = 3,
      ncol = 6
    ) +
    scale_x_continuous(breaks = seq(1, 14, 2)) +
    scale_y_continuous(
      limits = c(0, 8),
      breaks = seq(0, 8, 2)
    ) +
    xlab("Day") +
    ylab(expression("MIC (\u03BCg/mL)")) +
    theme_cowplot() +
    theme(panel.grid.major = element_line(
      color = "#3a3a3a",
      linewidth = 0.2
    )) +
    theme(plot.background = element_rect(fill = "#FFFFFF"))

  if (combined == FALSE) {
    return(vanc_plot)
  } else {
    # Combine the experimental schematic illustration and vancomycin plot
    # Imports PDF experimental schema image into R and adds it above the
    # vancomycin_MICs_plot
    schema <- image_read_pdf("Figures/experimental_schema.pdf",
      density = 300
    )

    schema_grob <- rasterGrob(as.raster(schema),
      interpolate = TRUE
    )

    combined_plot <- plot_grid(schema_grob,
      vanc_plot,
      ncol = 1,
      rel_heights = c(0.3, 0.7),
      labels = c("A", "B"),
      label_size = 16
    )

    combined_plot <- combined_plot +
      theme(plot.background = element_rect(fill = "#FFFFFF"))

    return(combined_plot)
  }
}
```
<br>

```{r Plot vancomycin MIC data}
vanc_plot <- generate_vanc_plot(vancomycin_MICs)

vanc_plot
```
<br>

```{r Generate plot with experimental schema}
schema_vanc_plot <- generate_vanc_plot(vancomycin_MICs,
  combined = TRUE
)

schema_vanc_plot
```
<br>

```{r Save plot of vancomycin MIC data}
ggsave("vancomycin_plot.png",
  vanc_plot,
  path = "Figures",
  device = "png",
  width = 14,
  height = 8,
  units = "in"
)
```
<br>

```{r Save plot of vancomycin MIC data with experimental schema}
ggsave("schema_vanc_plot.png",
  schema_vanc_plot,
  path = "Figures",
  device = "png",
  width = 14,
  height = 8,
  units = "in"
)
```
<br>

***

# Collateral responses analysis

Load data

```{r Collateral drug MIC data}
# Data report the MICs of 8 antibiotics -- cefazolin, clindamycin, daptomycin,
# gentamycin, meropenem, nafcillin, trimethoprim-sulfamethoxazole, and
# vancomycin -- on the MSSA ancestral clone ATCC 29213 and evolved
# vancomycin-intermediate populations.

collateral_MICs <- read_csv("data/collateral_drug_MICs.csv",
  show_col_types = FALSE
)
```


```{r Function that estimates collateral response values, messages = FALSE}
estimate_CR <- function(dat) {
  # This function takes a data frame of MICs and calculates drug response
  # values for each evolved population relative to the ancestral population.
  ancestor_frame <- dat %>%
    filter(population == "Ancestor")

  evolved_frame <- dat %>%
    filter(population != "Ancestor")

  CR_values <- c()

  # This code chunk iterates through the unique paired_IDs in the
  # evolved_frame and filters the ancestor_frame and evolved_frame for the
  # paired_ID. The MIC values for the ancestor and evolved populations are
  # then used to calculate the collateral response value for each evolved
  # population. The collateral response values are then appended to the
  # CR_values vector.
  for (i in unique(evolved_frame$paired_ID)) {
    ancestor_value <- ancestor_frame %>%
      filter(paired_ID == i) %>%
      pull(MIC)

    evolved_subframe <- evolved_frame %>%
      filter(paired_ID == i)

    for (j in 1:nrow(evolved_subframe)) {
      evolved_value <- evolved_subframe[j, ] %>%
        pull(MIC)

      CR_values <- c(CR_values, log2(evolved_value / ancestor_value))
    }
  }

  # Creates a tibble of collateral response values
  CR_col <- tibble(CR = CR_values)
  CR_df <- bind_cols(evolved_frame, CR_col)

  # Removed paired_ID column from CR_df
  CR_df <- CR_df %>%
    select(-paired_ID)

  return(CR_df)
}
```
<br>

```{r Estimate collateral response values, messages = FALSE}
collateral_responses <- estimate_CR(collateral_MICs)

head(collateral_responses)
```
<br>

```{r Function that estimates summary stats of CR values, messages = FALSE}
# Summary statistics function
perform_sum_stats <- function(dat) {
  summary_stats <- dat %>%
    group_by(antibiotic, population) %>%
    summarize(
      mean_MIC = mean(MIC),
      mean_CR = mean(CR)
    ) %>%
    filter(antibiotic != "VAN")

  # Create a new column called "state" in the summary_stats data frame.
  # Populate this column with "-1", "0", or " 1" if the mean collateral
  # response value is less than, equal to, or greater than 0.
  summary_stats$state <- ifelse(summary_stats$mean_CR < 0,
    yes = "-1",
    no = ifelse(summary_stats$mean_CR == 0,
      yes = "0",
      no = "1"
    )
  )

  return(summary_stats)
}
```
<br>

```{r Calculate summary statistics of CR values, messages = FALSE}
summary_stats <- perform_sum_stats(collateral_responses)

head(summary_stats)
```
<br>

We used Mann-Whitney U tests to compare the collateral responsevalues of
the evolved populations to the ancestral population for each antibiotic.

```{r Function that performs Mann Whitney U tests}
perform_MW_test <- function(dat, ab) {
  ancestor_df <- dat %>%
    filter(population == "Ancestor" & antibiotic == ab) %>%
    select(-replicate, -paired_ID)

  evolved_df <- dat %>%
    filter(population != "Ancestor" & antibiotic == ab) %>%
    group_by(population) %>%
    # Calculate the mean MIC for each evolved population
    summarize(MIC = mean(MIC)) %>%
    # Relabel all entries in the "population" column to "Evolved"
    mutate(population = ifelse(population == "Ancestor",
      yes = "Ancestor",
      no = "Evolved"
    ))

  # Bind the antibiotic_df and ancestor_df objects
  bound_df <- bind_rows(evolved_df, ancestor_df)

  # Perform Mann Whitney U test comparing the MICs of the ancestor and
  # evolved populations
  MW_test <- wilcox.test(MIC ~ population,
    data = bound_df,
    exact = FALSE,
    correct = FALSE,
    alternative = "two.sided"
  )

  return(MW_test)
}
```
<br>

```{r Perform Mann Whitney U tests}
mw_stats <- map(
  unique(collateral_MICs$antibiotic),
  ~ perform_MW_test(collateral_MICs,
    ab = .x
  )
)

names(mw_stats) <- unique(collateral_MICs$antibiotic)

list(mw_stats)
```
<br>

```{r Function that plots collateral response values}
generate_CR_plot <- function(dat) {
  # Factor antibiotics by order of decreasing median collateral response
  antibiotic_factor <- dat %>%
    group_by(antibiotic) %>%
    summarize(med = median(mean_CR)) %>%
    arrange(desc(med)) %>%
    pull(antibiotic)

  dat$antibiotic <- factor(dat$antibiotic,
    levels = as.factor(antibiotic_factor)
  )

  CR_boxplot <- dat %>%
    ggplot(aes(
      x = antibiotic,
      y = mean_CR,
      fill = antibiotic
    )) +
    geom_boxplot() +
    geom_jitter(
      shape = 16,
      position = position_jitter(0.2),
      size = 2
    ) +
    xlab("Antibiotic") +
    ylab(expression(
      "Log"[2] ~ "MIC"["evolved"] ~ "-" ~ "Log"[2] ~ "MIC"["ancestor"]
    )) +
    scale_fill_manual(
      values = c(
        "#D09384", "#E4BDB2", "#F7E8E6", "#FFFFFF",
        "#F3EEF9", "#E3DAF3", "#D4C3EA", "#CAB7E6"
      )
    ) +
    theme_cowplot() +
    theme(
      legend.position = "none",
      plot.background = element_rect(fill = "white")
    )
}
```
<br>

```{r Plot collateral response values}
CR_boxplot <- generate_CR_plot(summary_stats)

CR_boxplot
```
<br>

```{r Save plot of collateral response values}
ggsave("CR_boxplot.png",
  CR_boxplot,
  path = "Figures",
  device = "png",
  width = 10,
  height = 8,
  units = "in"
)
```
<br>

```{r Function that calculates probability of exhibiting a drug response}
calculate_likelihood <- function(dat) {
  prob_vec <- c()

  # For each antibiotic in the summary_stats data frame, we calculate the
  # median collateral response values using the mean_CR column.
  median_df <- dat %>%
    group_by(antibiotic) %>%
    summarize(median = median(mean_CR))

  # We iterate through the unique antibiotics in the summary_stats
  # data frame and calculates the probability of a population exhibiting a drug
  # response. This probability is defined as the fraction of evolved populations
  # that have a collateral response value that has the same sign as the median
  # collateral response value. The probability of a population exhibiting a drug
  # response is calculated for each antibiotic and stored in the prob_vec vector
  for (i in unique(dat$antibiotic)) {
    L <- 0

    ab_frame <- dat %>%
      filter(antibiotic == i)

    med_CR <- ab_frame %>%
      pull(mean_CR) %>%
      median()

    for (j in ab_frame$mean_CR) {
      if (sign(j) == sign(med_CR)) {
        # We initialize a counter (L) at zero and increment it by a fraction
        # (1 / total number of populations) for each mean_CR value that is in
        # concordance with the median.
        L <- L + (1 / nrow(ab_frame))
      }
    }

    prob_vec <- c(prob_vec, L)
  }

  probability_df <- tibble(
    antibiotic = unique(dat$antibiotic),
    probability = prob_vec
  )

  # Join the median_df and probability_df objects by antibiotic
  df <- median_df %>%
    left_join(probability_df, by = "antibiotic")

  return(df)
}
```
<br>

```{r Calculate population probability of exhibiting a drug response}
likelihood_df <- calculate_likelihood(summary_stats)

head(likelihood_df)
```
<br>

```{r Plot probability of drug response}
likelihood_plot <- likelihood_df %>%
  ggplot(aes(
    x = probability,
    y = median
  )) +
  geom_hline(
    yintercept = 0,
    linetype = "dashed"
  ) +
  geom_point(
    size = 5.5,
    fill = c(
      "#D09384", "#E4BDB2", "#FFFFFF",
      "#F3EEF9", "#E3DAF3", "#D4C3EA", "#CAB7E6"
    ),
    pch = 21
  ) +

  # Label each point with its corresponding antibiotic
  geom_text(aes(label = antibiotic),
    hjust = -0.5,
    vjust = -0.5,
    size = 5
  ) +
  scale_x_continuous(
    limits = c(0.5, 1),
    breaks = seq(0.5, 1, 0.1)
  ) +
  scale_y_continuous(
    limits = c(-1, 2),
    breaks = seq(-1, 2, 0.5)
  ) +
  xlab("Likelihood of drug response") +
  ylab(
    expression("Log"[2] ~ "MIC"["evolved"]
    ~ "-" ~ "Log"[2] ~ "MIC"["ancestor"])
  ) +
  theme_cowplot() +
  theme(
    legend.position = "none",
    plot.background = element_rect(fill = "white")
  )

likelihood_plot
```
<br>

```{r Save plot of probability of drug response}
ggsave("likelihood_plot.png",
  likelihood_plot,
  path = "Figures",
  device = "png",
  width = 10,
  height = 8,
  units = "in"
)
```
<br>

***

# Genomic evolution

We used the *breseq* pipeline (v0.39.0) to identify mutations in the
control and experimental populations. The control lines evolved in
TSB medium without vancomycin, whereas the experimental lines evolved
in TSB medium supplemented with vancomycin. We concatenated the breseq
output files for each control and experimental population, labeled "C1"
through "C87" and "E1" through "E18", respectively. We also manually
curated the breseq output files to ensure that the mutations were correctly
identified and annotated. Lastly, following Deatherage et al. 2017 and
Card et al. 2021, for each mutation, we determined whether it "qualified"
for further analyses based on the following criteria:

1. It was a nonsynonymous point mutation or small indel in a single gene.
2. An intergenic mutation within 150 bp upstream of the start of a gene.
3. A large deletion if at least one of the affected genes was also found
to be mutated in another population.

Conversely, we excluded mutations from further analyses if they were:

1. Synonymous (does not affect the resulting amino acid sequence).
2. Multicopy elements (e.g., ribosomal RNA operons) that may result from
gene conversions but cannot be fully resolved using short-read sequencing data.

We also include all mutations that were at or above 5% frequency in the
population given that each population had an average read coverage depth of
>100X. A total of *227* mutations qualify based upon these criteria in the
vancomycin-evolved lines and *3585* mutations qualify in the control lines.

```{r Reads in mutation data, messages = FALSE}
mutations <- read_csv("data/mutations.csv",
  show_col_types = FALSE
)

# Filter the mutations object by the presence of qualifying mutations.
qualifying_mutations <- mutations %>%
  filter(qualifying_mutation == "yes")
```
<br>

This code chunk generates a binary matrix representing the qualifying
gene-level mutations in each evolutionary replicate. We use this matrix for
subsequent analyses.

```{r Generation of binary data frame of mutations}
generate_binary_matrix <- function(dat, omit = NULL) {
  # Creates a vector of the unique sample designations in the
  # qualifying_mutations object
  populations <- unique(dat$population)

  # Sorts the genes with qualifying_mutations by number of hits and
  # creates a vector
  genes <- dat %>%
    group_by(gene) %>%
    summarize(hits = n()) %>%
    arrange(desc(hits)) %>%
    pull(gene)

  # If the omit argument is not NULL, then the genes vector is filtered to
  # exclude the genes in the omit vector (used in downstream analysis)
  if (!is.null(omit)) {
    genes <- genes[!genes %in% omit]
  }

  # Creates a binary matrix with the number of rows equal to the number of
  # samples, and the number of columns equal to the number of genes in the
  # mutations object and populates the matrix with 0s.
  binary_mat <- matrix(0,
    nrow = length(populations),
    ncol = length(genes)
  )

  rownames(binary_mat) <- populations
  colnames(binary_mat) <- genes

  # Use the qualifying_mutations object to populate the binary_mat object.
  # This code chunk iterates through the qualifying mutations data frame
  # and populates the binary_mat object with a 1 if a mutation is present in a
  # given gene / lineage combination
  for (i in seq_len(nrow(dat))) {
    binary_mat[dat$population[i], dat$gene[i]] <- 1
  }

  return(binary_mat)
}
```
<br>

```{r Generate binary matrix}
binary_matrix <- generate_binary_matrix(qualifying_mutations)
```
<br>

We quantified the extent of genomic parallelism in the vancomycin-adapted lines
and in the control lines by calculating Dice's Similarity Coefficient (*S*)
for each population pair, where:

$$S = \frac{2|X \cap Y|}{|X|+|Y|}$$

$|X|$ and $|Y|$ represent the number of genes with qualifying mutations in each
population, and $|X \cap Y|$ is the number of mutated genes in common between
them. *S* therefore ranges from 0, when the pair share no mutations in common,
to 1, when both have mutations in exactly the same set of genes
(Deatherage et al., 2017; Card et al. 2021).

By default, the perform_genomic_analysis function will output a list containing
two elements:

1. Average pairwise similarity within the vancomycin-treated group
($\overline{S}_v \approx 0.0625$) and within the control group
($\overline{S}_c \approx 0.0290$). In other words, two populations that
independently evolved under vancomycin selection had, on average, *6.25%* of
their mutated genes in common, whereas those that evolved under identical
conditions, but without vancomycin, shared on average only *2.90%* of their
mutated genes.

2. The difference between these averages
($\overline{S}_v - \overline{S}_c \approx 0.0335$) represents how much greater
(on average) the vancomycin-adapted lines' similarity is relative to the control
lines. We expect that the vancomycin-adapted lines will exhibit a higher degree
of genomic parallelism than the control lines.

To evaluate the statistical significance of the observed difference between
the vancomycin-adapted and control lines, we generated a bootsrap distribution
assuming no difference between the two groups. We pooled all data into one
combined dataset, then repeatedly sampled (with replacement) two new subsets of
the same sizes as the original groups. For each bootstrap iteration
(10,000 total), we calculated the difference in mean similarities between the
subsets. We then compared our observed difference to the distribution of
differences generated by the bootstrap analysis. The *p*-value is the proportion
of bootstrap iterations where the difference was greater than or equal to the
observed difference.

```{r Genomic Analysis function}
perform_genomic_analysis <- function(dat, bootstrap = FALSE) {
  if (bootstrap == TRUE) {
    # Sample the lines, with replacement, 105 times.
    dat <- dat[sample(nrow(dat), replace = TRUE), ]

    # Rename the rows of the sampled dataset - some rows are sampled
    # more than once. By default R will append a number to the end of the
    # row name to differentiate between these rows. These rows are then
    # treated separately in the downstream analyses. This code chunk renames
    # the rows to C1, C2, ..., C87 and E1, E2, ..., E18 etc
    # to avoid this issue.
    rownames(dat) <- c(paste0("C", 1:87), paste0("E", 1:18))
  }

  # Computes Dice's Similarity Coefficient for all possible replicate pairs.
  # Populates a matrix with these values
  pwise_simil_matrix <- as.matrix(
    simil(dat,
      method = "Dice",
      by_rows = TRUE,
      upper = TRUE,
      diag = TRUE
    )
  )

  # All values above (and including) the matrix diagonal are converted to
  # NA to ease downstream wrangling
  diag(pwise_simil_matrix) <- NA
  pwise_simil_matrix[upper.tri(pwise_simil_matrix)] <- NA

  # Converted matrix to a data frame for analysis. NA values above
  # (and including) the matrix diagonal are incorporated into the newly
  # formed data frame. This piece of code drops the rows containing NA,
  # effectively retaining only those values *below* the matrix diagonal
  pwise_simil_df <- as.data.frame.table(pwise_simil_matrix,
    responseName = "value"
  ) %>%
    rename(
      population_1 = Var1,
      population_2 = Var2
    ) %>%
    drop_na() %>%
    # Remove numbers at the end of the population names
    mutate(
      population_1 = str_remove_all(population_1, "\\d"),
      population_2 = str_remove_all(population_2, "\\d")
    )


  ## First part of the summary output if not performing a bootstrap analysis

  # Computes the mean pairwise similarity within the vancomycin group
  # and control group
  avg_similarity_df <- pwise_simil_df %>%
    group_by(population_1, population_2) %>%
    summarize(avg = mean(value)) %>%
    as_tibble() %>%
    filter(population_1 == population_2)


  ## Second part of the summary output if not performing a subsequent
  ## bootstrap analysis

  difference <- avg_similarity_df[2, 3] - avg_similarity_df[1, 3]

  names(difference) <- "difference"

  # If the perform_genomic_analysis function is told to perform a bootstrap
  # analysis, then only the difference is returned. The significance of this
  # statistic is then calculated.
  if (bootstrap == FALSE) {
    results <- list(avg_similarity_df, difference)
  } else {
    # Makes a data frame with three columns:
    # 1. Bootstrap average pairwise similarity - control populations
    # 2. Bootstrap average pairwise similarity - experimental populations
    # 3. Difference between these bootstrap means
    results <- avg_similarity_df %>%
      mutate(treatment = ifelse(population_1 == "C",
        yes = "control",
        no = "experimental"
      )) %>%
      select(-population_1, -population_2) %>%
      pivot_wider(
        names_from = treatment,
        values_from = avg
      ) %>%
      mutate(difference = experimental - control)
  }

  return(results)
}
```
<br>

```{r Calculate genomic similarity and differences between groups}
similarity <- perform_genomic_analysis(binary_matrix,
  bootstrap = FALSE
)

similarity
```
<br>

```{r Fuction that performs bootstrap analysis}
perform_bootstrap_analysis <- function(dat) {
  # This function iterates through the perform_genomic_analysis function
  # 10,000 times and stores the results in a data frame. The function then
  # returns the data frame.

  # Detect number of cores and register parallel backend
  cl <- makeCluster(detectCores() - 1)
  registerDoParallel(cl)

  # Bootstrap distributions of the average pairwise similarity in each treatment
  results <- foreach(
    i = 1:10000,
    .combine = rbind,
    .packages = c("tidyverse", "proxy"),
    .export = "perform_genomic_analysis"
  ) %dopar% {
    perform_genomic_analysis(dat,
      bootstrap = TRUE
    )
  }

  stopCluster(cl)

  return(results)
}
```
<br>

```{r Perform bootstrap analysis}
bootstrap_results <- perform_bootstrap_analysis(binary_matrix)

significance <- sum(
  bootstrap_results$difference >= similarity[[2]]$difference
) / 10000
```
<br>

```{r Plot the bootstrap distribution of differences}
bootstrap_plot <- bootstrap_results %>%
  ggplot(aes(x = difference)) +
  geom_histogram(
    binwidth = 0.001,
    fill = "#7C8DA2"
  ) +
  geom_vline(
    xintercept = similarity[[2]]$difference,
    linetype = "dashed",
    color = "#D09384"
  ) +
  scale_y_continuous(breaks = seq(0, 600, 100)) +
  annotate("text",
    x = similarity[[2]]$difference + 0.012,
    y = 550,
    label = "Observed difference",
    color = "#D09384"
  ) +
  annotate("text",
    x = similarity[[2]]$difference + 0.012,
    y = 500,
    label = paste(
      "p =",
      round(significance, 4)
    ),
    color = "#D09384"
  ) +
  xlab("Difference in mean similarity") +
  ylab("Frequency") +
  theme_cowplot() +
  theme(legend.position = "none")

bootstrap_plot
```
<br>

```{r Save bootstrap plot}
ggsave("bootstrap_plot.pdf",
  bootstrap_plot,
  path = "Figures",
  device = "pdf",
  width = 10,
  height = 8,
  units = "in"
)
```
<br>

The bootstrap analysis showed that the vancomycin-adapted lines differ
significantly in their genetic responses relative to the control lines.
However, this analysis does not specify which genes are significantly
associated with each treatment. To identify these genes, we perform
a multivariate logistic regression with the treatment status (i.e., vancomycin
or no vancomycin) as the outcome and the presence or absence of a mutation in
each gene as the predictors. We only include in our model qualifying mutations
that occur in at least three populations resulting in 446 gene-level mutations.
Because there are only 18 outcome events (i.e., populations that evolved under
vancomycin selection) and 446 predictors, we use L1 penalization on model
coefficients to avoid overfitting.

```{r Logistic regression function}
perform_regression <- function(dat) {
  # Transpose the binary_mat data frame
  regression_df <- dat %>%
    t() %>%
    as.data.frame() %>%
    rownames_to_column(var = "treatment") %>%
    # Remove numbers at the end of the population names
    mutate(treatment = str_remove_all(treatment, "\\d")) %>%
    # If the value in the treatment column is E, replace it with 1.
    # Otherwise, replace it with 0
    mutate(treatment = ifelse(treatment == "E", 1, 0)) %>%
    # Remove columns in which the sum of all values is less than 3
    select_if(~ sum(.) >= 3)

  # Multivariate logistic regression routine over 100 iterations to select
  # more robust non-zero coefficients

  # Detect number of cores and register parallel backend
  cl <- makeCluster(detectCores() - 1)
  registerDoParallel(cl)

  coef_count <- foreach(
    i = 1:100,
    .combine = "+",
    .multicombine = FALSE,
    .init = numeric(ncol(regression_df[-1])),
    .packages = c("glmnet")
  ) %dopar% {
    cvfit <- cv.glmnet(
      as.matrix(regression_df[, -1]),
      as.factor(regression_df$treatment),
      family = binomial(link = "probit"),
      gamma = seq(0, 1, 0.1),
      nfolds = 3,
      maxit = 1e7,
      relax = TRUE
    )

    fit_res <- glmnet(
      as.matrix(regression_df[, -1]),
      as.factor(regression_df$treatment),
      family = binomial(link = "probit"),
      lambda = cvfit$relaxed$lambda.min,
      alpha = cvfit$relaxed$gamma.min
    )

    coef_res <- as.matrix(fit_res$beta)
    idx <- which(coef_res != 0)

    # Create a temporary vector for this iteration only
    tmp_count <- numeric(ncol(regression_df[-1]))
    tmp_count[idx] <- 1

    return(tmp_count)
  }

  stopCluster(cl)

  idx <- which(coef_count >= 20)

  # Select columns of the regression_df data frame based on the idx vector
  # (i.e., these coefficients were associated with treatment status in at least
  # 20 of the 100 iterations)
  coef_matrix <- as.matrix(regression_df[-1][, idx])
  rownames(coef_matrix) <- colnames(dat)

  # Subset the matrix to include only the experimental populations
  exp_coef_matrix <- coef_matrix[88:105, ]

  # Order the columns in the exp_coef_matrix by the sum of values in each
  colnames_exp <- exp_coef_matrix[, order(colSums(exp_coef_matrix),
    decreasing = TRUE
  )] %>%
    colnames()

  # Order the columns in the coef_matrix object by colnames_exp
  coef_matrix <- coef_matrix[, colnames_exp]

  return(coef_matrix)
}
```
<br>

```{r Perform logistic regression}
coef_matrix <- perform_regression(binary_matrix)

head(coef_matrix)
```
<br>

```{r Plot mutation heat map}
mutation_heatmap <- pheatmap(coef_matrix,
  cluster_rows = FALSE,
  cluster_cols = FALSE,
  color = colorRampPalette(c("#E7E5DF", "#606B91"))(100),
  gaps_row = c(87, 87, 87),
  legend = FALSE,
  angle_col = 0,
  border_color = "#FFFFFF"
)
```
<br>

```{r Save mutation heat map}
ggsave("mutation_heatmap.png",
  mutation_heatmap,
  path = "Figures",
  device = "png",
  width = 10,
  height = 15,
  units = "in"
)
```
<br>

Twenty gene-level mutations were significantly associated with vancomycin
selection and 13 with the control treatment. The most common mutations in the
